Feature/plantuml diagram

doug-harriman opened the pull request:
Added ability to generate PlantUML and Graphviz dot state chart diagrams.

pgularski commented on test/test_plantuml.py:
```diff
@@ -0,0 +1,22 @@
+from test_simple_on_off import sm
```
I'm not sure the generated staruml covers all capabilities of the library. Ie. Can it generate a multi level HSM? What about historical states? Can you add tests to cover these scenarios?

pgularski reviewed (commented):
Overall I really like the idea. I was thinking about it myself. I didn't add such a code though for several reasons. Mainly because:
- I'd rather have some external library, a wrapper around the main pysm library, in order to limit the main library size - so that it keeps working in space and memory limited embedded systems. To facilitate it, I'd add some hooks to the main library and use them in extending libraries.
- Complex HSMs tend to be very messy when presented in a chart - but after some consideration I'd say that it's still better to have a possibility to generate a chart than not, so this point isn't valid.

Nonetheless, I like your changes and these seem to be small enough to work with embedded systems . Just make sure you address things in my code comments.

Thanks!

doug-harriman commented:
@pgularski - Before I address your points above, let me first apologize.  I created this pull request by accident.  I created a fork of your project, and meant to create pull request back to the master branch on that fork, not your project.  I'm happy to merge back to your fork if you would like me to, please respond to this to confirm that is your wish.  I made a lot of additional progress after this pull request was created.   Some things to note about my fork:

* My intent is to use this library for a large machine that is PC hosted.  As such, I'm not as concerned about uPy support.
* My main concerns are adding two key features (both completed in my fork): 
   * Calling all state entry function upon initialization.  This supports clean initialization without requiring an additional state & message dispatch.
   * Historical states.  I know I'm going to need a pause/resume functionality, so the ability to re-enter a "processing" state that has sub-states for individual process steps, and to jump back to the next step in that process is critical.
* I'm adding tools to help with debugging of complex state machines.  
   * Generating the diagram is the first step, as it allows easy communications of the design intent, as well as finding errors visually.  They definitely can get big, but they're still useful.
   * I've added the ability to highly the current state, and plan on adding more of that.  Later I'm hoping to add log replay graphically showing transitions and states taken.
   * I'm going to add state and transition visit counts, so I can easily identify statechart coverage in testing, and highlight those that were not visitied.  

Note that I've also implemented a github workflow with flake8 linting and running the pytest tests, so the formatting issues have all been resolved.

doug-harriman commented:
@pgularski - Before I address your points above, let me first apologize.  I created this pull request by accident.  I created a fork of your project, and meant to create pull request back to the master branch on that fork, not your project.  I'm happy to merge back to your fork if you would like me to, please respond to this to confirm that is your wish.  I made a lot of additional progress after this pull request was created.   Some things to note about my fork:

* My intent is to use this library for a large machine that is PC hosted.  As such, I'm not as concerned about uPy support.
* My main concerns are adding two key features (both completed in my fork): 
   * Calling all state entry function upon initialization.  This supports clean initialization without requiring and additional state & message dispatch.
   * Historical states.  I know I'm going to need a pause/resume functionality, so the ability to re-enter a "processing" state that has sub-states for individual process steps, and to jump back to the next step in that process is critical.
* I'm adding tools to help with debugging of complex state machines.  
   * Generating the diagram is the first step, as it allows easy communications of the design intent, as well as finding errors visually.
